OPHAKTKEH-26 Dockerized services for dev usage

pkoivisto opened the pull request.

pkoivisto committed:
Add Dockerfile for the backend

pkoivisto committed:
Add Dockerfile for the React app

pkoivisto committed:
Add docker-compose.yml to wire services together

pkoivisto committed:
Simplify frontend setup

pkoivisto committed:
Setup and provide postgres as a service with Docker Compose

terova commented:
Olisiko tarpeellista READMEen jotakin kirjoittaa?

pkoivisto commented:
> Olisiko tarpeellista READMEen jotakin kirjoittaa?

Hyvä huomio! Ainakin noi aiemmat ohjeet postgres-kuvan ajamisesta ovat tämän jälkeen turhat. Päivitän ohjeet.

mikhuttu commented on db/Dockerfile:
```diff
@@ -1,4 +1,2 @@
-ENV POSTGRES_PASSWORD postgres
-COPY /db/*.sql /docker-entrypoint-initdb.d/
\ No newline at end of file
+COPY *.sql /docker-entrypoint-initdb.d/
```
Skriptin `create_db_sql.sh` voinee poistaa?

pkoivisto replied:
Sitähän ei siis tuossa kopioidu mukaan (koska `.sql`) mutta tarkoititko että poistettaisiin kokonaan versionhallinnasta?

terova replied:
Eihän se tuolla versionhallinnassa mitään haittaa, on valmiina sitten jos pitää sql tiedostoja muuttaa, niin ei tarvitse uudestaan muistella mitenkä ne otetaankaan.

pkoivisto replied:
Jätetään versionhallintaan.

mikhuttu commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM maven:3.8.3-openjdk-17-slim
```
3.8.4:ään voitanee halutessa vaihtaa mutta tämä ok

pkoivisto replied:
Päivitetään uudempaan.

mikhuttu commented on src/main/reactjs/Dockerfile:
```diff
@@ -0,0 +1,13 @@
+FROM node:14-slim
```
Taitaa vastata `node:14.18.1-slim` imagea? Vois mahd. vaihtaa nimen siihen.

pkoivisto replied:
Muutettu.

mikhuttu reviewed (commented):
Hyvältä vaikuttaa kunhan frontendin ajamiseen liittyvät virheet saa fiksattua. Itselle tulee käynnistyksen aikana seuraavanlaista lokia (testasin nykyisen version päällä, jossa devin muutokset mukana):

```
backend_1   | [INFO] Running 'npm install' in /app/src/main/reactjs
frontend_1  | WARNING in /app/reactjs/src/configs/i18n.ts
frontend_1  |   46:10  warning  Caution: `i18n` also has a named export `use`. Check if you meant to write `import {use} from 'i18next'` instead                        import/no-named-as-default-member
frontend_1  |   67:10  warning  Caution: `i18n` also has a named export `changeLanguage`. Check if you meant to write `import {changeLanguage} from 'i18next'` instead  import/no-named-as-default-member
frontend_1  |
frontend_1  | ✖ 2 problems (0 errors, 2 warnings)
frontend_1  |
frontend_1  | webpack compiled with 1 warning
```

pkoivisto committed:
Pin npm used within the frontend container to correct version

pkoivisto commented:
> Hyvältä vaikuttaa kunhan frontendin ajamiseen liittyvät virheet saa fiksattua. Itselle tulee käynnistyksen aikana seuraavanlaista lokia (testasin nykyisen version päällä, jossa devin muutokset mukana):
> 
> ```
> backend_1   | [INFO] Running 'npm install' in /app/src/main/reactjs
> frontend_1  | WARNING in /app/reactjs/src/configs/i18n.ts
> frontend_1  |   46:10  warning  Caution: `i18n` also has a named export `use`. Check if you meant to write `import {use} from 'i18next'` instead                        import/no-named-as-default-member
> frontend_1  |   67:10  warning  Caution: `i18n` also has a named export `changeLanguage`. Check if you meant to write `import {changeLanguage} from 'i18next'` instead  import/no-named-as-default-member
> frontend_1  |
> frontend_1  | ✖ 2 problems (0 errors, 2 warnings)
> frontend_1  |
> frontend_1  | webpack compiled with 1 warning
> ```

Korjattu uudemmassa versiossa, npm-version pinnaaminen kontin sisällä auttoi.

pkoivisto committed:
Fix base image

pkoivisto committed:
Review feedback: use newer Maven image

pkoivisto committed:
Setup development time proxy to transparently route API requests from frontend to backend.

p3j4m commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+FROM maven:3.8.4-openjdk-17-slim
```
Tässä voisi käyttää imagea: eclipse-temurin:17-jdk-alpine

pkoivisto replied:
Miten olisi `eclipse-temurin:17-jdk`? alpine-versiosta ei ole buildattu arm64:lle versiota, joten ei suoraan toimi mun M1:llä.

p3j4m replied:
Se on myös ok.

p3j4m commented on Dockerfile:
```diff
@@ -0,0 +1,8 @@
+WORKDIR /app
+
+ADD . ./
+RUN mvn clean install -DskipTests
```
Käytetään maven wrapperia ja skipataan frontendin buildaukset:

RUN ./mvnw clean install -DskipTests -Dskip.npm

CMD ["./mvnw", "spring-boot:run", "-Dskip.npm"] 


Tämä tarvitsee PR: https://github.com/Opetushallitus/akt/pull/24

p3j4m commented on src/main/reactjs/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+FROM node:14.18.1-slim
```
Pitäisikö käyttää alpine imagea?

pkoivisto replied:
Sopii.

p3j4m commented on src/main/reactjs/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+
+COPY . ./
+
+RUN npm run build
```
Mihin tätä tarvitaan? npm startin pitäisi onnistua ilman tätäkin.

pkoivisto replied:
Joo, eipä tarvitakaan. Poistan.

p3j4m commented on src/main/reactjs/Dockerfile:
```diff
@@ -0,0 +1,16 @@
+
+# Pin npm version globally so that it won't get overwritten 
+# by the package.json copied in later steps. 
+RUN npm i -g --silent ${npmversion}
```
Jos käyttää imagea node:14.18.1-slim tai alpine ei erillistä npm-asennusta tarvita enää. Minulla toimii alla oleva versio ongelmitta:

FROM node:14.18.1-alpine
WORKDIR /app/reactjs

COPY package*.json ./
RUN npm ci --silent

COPY . ./

CMD ["npm", "start"]

pkoivisto replied:
Vaikuttaa tosiaan toimivan. Aika montaa eri varianttia kokeilin ennen kuin sain toimimaan siten että toi Mikan mainitsema varoitus poistui. Tuolla node-imagella npm:stä asentuu versio `6.14.15`, mutta ehkäpä se ei haittaa kun ajetaan `npm ci`.

pkoivisto replied:
Joka tapauksessa otan ton sun version käyttöön, niin yksinkertaistuu tää setuppi.

pkoivisto committed:
Update README.md

pkoivisto committed:
Review feedback: simplify frontend Dockerfile

pkoivisto committed:
Review feedback: use maven-wrapper + eclipse-temurin JDK image, don't run npm install for backend container

pkoivisto committed:
Remove unnecessary links block

pkoivisto committed:
Pass address to proxy API requests to as an environment variable to webpack.

Add additional npm script 'start:docker-compose' for launching the frontend app with the correct proxy address when run in the container.
